### ISSUE TITLE
[$500] IOU - Amount page reloaded when navigate back from final confirmation and click next btn

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#30060]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.87-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause-Internal Team
**Slack conversation:** @

## Action Performed:
Precondition: user should be signed in
1. Open app
2. Click on FAB
3. Select the Request Money
4. Select Manual tab
5. Write any amount and click Next button
6. Select any attendees (two or more)
7. Click back button twice in a row
8. Click next button

## Expected Result:
User should be navigated on the select attendees page

## Actual Result:
The same page reloaded with default amount (the default amount is zero)

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/115492554/d16e8fd5-45b4-4728-951d-4e73ed07ffcb

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/115492554/986c4fb4-6142-4319-aefd-77c854117b68

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~014945c77a33a9f852</li>
        <li>Upwork Job ID: 1715293865929134080</li>
        <li>Last Price Increase: 2023-11-03</li>
<li>Automatic offers: </li>
<ul>
<li>jjcoffee | Reviewer | 27561168</li>
<li>graylewis | Contributor | 27561170</li>
</ul></li>
    </ul>
</details>

**Comment by graylewis** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
When using the IOU manual request flow, attempting to go back to the requestMoneyAmount page and then moving back forward breaks the flow and resets the request

### What is the root cause of that problem?
Once you reach the confirmation page, the route of the app changes from /request to /split

```
On the participants page:
http://localhost:8082/request/new/participants/

On the confirmation page:
http://localhost:8082/split/new/confirmation/
```

Because the ID of a request is generated from the route, the safety measure preventing two separate requests on different tabs from interfering with each other is triggered, due to the conflict between IDs. This causes the request to be reset.

### What changes do you think we should make in order to solve the problem?
Currently, hitting the back button in this flow triggers a typical back action. We can maintain the ID properly by enforcing a fallback route rather than allowing the browser to determine which route to navigate to. 

The navigateBack function for the confirmation page before (note that the Navigation.goBack() function provides a fallback, but doesn't supply the optional second parameter to enforce navigation to the fallback):
```
    const navigateBack = () => {
        let fallback;
        if (reportID.current) {
            fallback = ROUTES.MONEY_REQUEST.getRoute(iouType.current, reportID.current);
        } else {
            fallback = ROUTES.MONEY_REQUEST_PARTICIPANTS.getRoute(iouType.current);
        }
        Navigation.goBack(fallback);
    };
```

The navigateBack function after (note the optional second parameter enforcing the fallback route in the Navigation.goBack call:
```
    const navigateBack = () => {
        let fallback;
        if (reportID.current) {
            fallback = ROUTES.MONEY_REQUEST.getRoute(iouType.current, reportID.current);
        } else {
            fallback = ROUTES.MONEY_REQUEST_PARTICIPANTS.getRoute(iouType.current);
        }
        Navigation.goBack(fallback, true);
    };
```

**Comment by FitseTLT** [comment]
This is related to #29972 and I recommend the same [solution](https://github.com/Expensify/App/issues/29972#issuecomment-1771305247) I gave there that will solve both of the issues.

**Comment by graylewis** [comment]
> This is related to #29972 and I recommend the same [solution](https://github.com/Expensify/App/issues/29972#issuecomment-1771305247) I gave there that will solve both of the issues.

I wouldn't say this solution for https://github.com/Expensify/App/issues/29972 actually solves this issue. It would stop the value from being reset but the rest of the stored information (participants, description, etc.) would be reset upon navigating back to the start of the user flow. Since the route in this case is supposed to act as the source of truth for the state of the app, it makes more sense to reflect the state of the app accurately in the route.

**Comment by FitseTLT** [comment]
> > This is related to #29972 and I recommend the same [solution](https://github.com/Expensify/App/issues/29972#issuecomment-1771305247) I gave there that will solve both of the issues.
> 
> I wouldn't say this solution for #29972 actually solves this issue. It would stop the request amount from being reset but the rest of the stored information (participants, description, etc.) would be reset upon navigating back to the start of the user flow. Since the route in this case is supposed to act as the source of truth for the state of the app, it makes more sense to reflect the state of the app accurately in the route.

We can also reset only `id` and other necessary props of the iou in here
https://github.com/Expensify/App/blob/eb5fec2c258e74a69bbed9d93100db4b30b265ec/src/libs/actions/IOU.js#L2863-L2871
But my logic is this function is called after amount is already input by the user (whether from manual, distance, ...) so resetting `iou.amount` because there is a mismatch b/n iouType and iou.id is not logical and that's why the app re-navigates to the amount page.

**Comment by FitseTLT** [comment]
@peterdbarkerUK @jjcoffee please check my [updated proposal](https://github.com/Expensify/App/issues/30060#issuecomment-1772606570)

**Comment by jjcoffee** [comment]
Reviewing tomorrow!

**Comment by jjcoffee** [comment]
@graylewis's [proposal](https://github.com/Expensify/App/issues/30060#issuecomment-1772484455) LGTM! The RCA is correct and the solution seems appropriate. Just a side note though, it's much easier to review your proposal if you provide links to the parts of code you're talking about (see @dukenv0307's [proposal](https://github.com/Expensify/App/issues/30060#issuecomment-1772523488) for inspiration!).

:ribbon::eyes::ribbon: C+ reviewed

**Comment by jjcoffee** [comment]
Hmm I think this is small enough that it would just be handled in the PR. I think I'm happy that @graylewis's [proposal](https://github.com/Expensify/App/issues/30060#issuecomment-1772484455) covered the RCA and the overall solution. Will defer to @roryabraham if he disagrees!

**Comment by peterdbarkerUK** [comment]
Assigned to @graylewis, with an extra point towards the proposal feedback [here](https://github.com/Expensify/App/issues/30060#issuecomment-1790503182), and the additional consideration to account for [here](https://github.com/Expensify/App/issues/30060#issuecomment-1790540683).

**Comment by jjcoffee** [comment]
Actually we might need to put this one on hold for https://github.com/Expensify/App/pull/28618 as this would be fixed there. I'm unsure how quickly that will be rolled out though, so depends on whether or not we want a fix in the meantime. cc @peterdbarkerUK

**Comment by graylewis** [comment]
@jjcoffee @peterdbarkerUK I'm a little confused. I'm already hired on upwork to handle the fix. Should I just go ahead and fix it or should I hold off?

**Comment by jjcoffee** [comment]
Let's wait on @peterdbarkerUK's response for now. Sorry for the confusion!

**Comment by peterdbarkerUK** [comment]
@graylewis I'm sorry for the confusion! Let's put this on hold for https://github.com/Expensify/App/pull/28618!

**Comment by roryabraham** [comment]
Still on HOLD

**Comment by paultsimura** [comment]
This looks very much like a duplicate of https://github.com/Expensify/App/issues/28129, which was created earlier and was on hold as well up until now.

**Comment by roryabraham** [comment]
Still on HOLD

**Comment by roryabraham** [comment]
Looks like this can come off HOLD

**Comment by jjcoffee** [comment]
No longer reproducible. Let's close this @peterdbarkerUK or @roryabraham
